### PR TITLE
[Peras 25] Conformance and property tests for voting committee implementations

### LINKED ISSUE #195: Implement committee selection schemes
Unless there already exists a library for that, we need to implement committee selection schemes to decide when stakeholders need to vote.

A good starting point would be `wFA^{IID}`, for simplicity, until we can replace it with a scheme that is more private and local, such as `wFA^{LS}` with a dedicated, non-trivial aggregation scheme.

See [this paper](https://eprint.iacr.org/2023/1273.pdf) for description of committee election schemes.

### LINKED ISSUE #210: Voting Committee selection logic

### PR DESCRIPTION
**Pull request by agustinmista**
This PR provides:

* Property tests for generic committee helpers (`VotesWithSameTarget`)
* Conformance tests for `WFALS` between the real implementation and:
  + the (Haskell) model, against randomly generated stake distributions,
  + the (Rust) prototype, against its precomputed outputs for the example stake distribution taken from mainnet.
* Property tests for both `WFALS` and `EveryoneVotes`, asserting that:
  + votes cast by eligible voters are successfully verifiable under the same voting committee, and
  + fake (illegal) votes are successfully rejected.

NOTES:

* The first two commits reorganize a bit the test suite (originally only containing WFALS tests), to share definitions and make room for `EveryoneVotes` tests.

### CODE REVIEW

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+      unNonPersistentCommitteeSize (nonPersistentCommitteeSize committee)
+
+    persistentRange =
+      (0, numPersistentSeats - 1)
+
+    nonPersistentRange =
+      (numPersistentSeats, numPersistentSeats + numNonPersistentSeats - 1)
+
+    -- This witness looks like a persistent member, but its seat index lies
+    -- outside the persistent members range (case 1).
```
Comment (dnadales) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs:378]: I wonder if we could hit an inverted `choose` range here when `numNonPersistentSeats == 0`? In that case `nonPersistentRange` becomes `(n, n - 1)` on `Word64`, and `genFakePersistentMemberWitness` calls `choose` on it — `randomR` does not guarantee anything when `lo > hi`.

Unlike the EveryoneVotes generator, which guards `genPoolWithZeroStakeWitness` behind a frequency-weight check, both WFALS alternatives always have weight 1.

What about something like the EveryoneVotes pattern — gate each alternative on whether its range is non-empty? Something like:

```haskell
frequency
  [ (if numNonPersistentSeats > 0 then 1 else 0, genFakePersistentMemberWitness)
  , (if numPersistentSeats > 0 then 1 else 0, genFakeNonPersistentMemberWitness)
  ]
```

(The second guard covers `persistentRange` being empty if `numPersistentSeats == 0`, though that may not happen in practice.)

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+                            witness
+                    case verifyVote committee vote of
+                      Left err ->
+                        counterexample
+                          ("vote verification failed: " <> show err)
+                          $ property False
+                      Right witness' ->
+                        counterexample
+                          "vote verificateion mismatch"
+                          $ eqWithShowCmp
```
Comment (dnadales) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs:153]: Tiny typo: `"verificateion"` → `"verification"`?

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS.hs [hunk 3]
```diff
@@ -1,36 +1,278 @@
...
-            (fromIntegral targetCommitteeSize)
-            (Map.map Model.rationalToStake stakeDistr)
-     in ( Map.size persistentSeats
+  -- NOTE: this is not a good tiebreaker for real-world use, but it is
+  -- sufficient here because the actual order of pools with the same stake does
+  -- not matter when computing the persistent vs. non-persistent seat /counts/.
+  tiebreaker =
+    compare
+
+  mkStakeDistr =
```
Comment (dnadales) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS.hs:68]: Tiny typo: `"strake"` → `"stake"`?

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/Utils.hs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+-- | Generate a positive stake value using a geometric distribution.
+genPositiveStake :: Gen LedgerStake
+genPositiveStake =
+  LedgerStake
+    . toRational
+    . (+ 1)
+    <$> geometric 0.25
+
+-- | Generate an alphanumeric string of length 8.
+-- | Generate a non-empty map of pools with crypto keys and varying stakes.
```
Comment (dnadales) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/Utils.hs:93]: Should we drop this first haddock line? It looks like it describes the internal `alphaNumString` helper rather than `genPools` — GHC will silently discard it since the next `-- |` wins.

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+                            , "but got these " <> show (length witnessesFromCert) <> " witnesses:"
+                            , intercalate "\n" (fmap showWitness witnessesFromCert)
+                            ]
+                        )
+                        $ cmpWitnesses originalWitnesses witnessesFromCert
+
+-- * WFALS helpers
+
+castVote ::
+  PrivateKey TestCrypto ->
```
Comment (dnadales) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs]: I notice `castVote` reimplements `forgeVote` / `implForgeVote` from scratch rather than calling it. Would it make sense to use `forgeVote` from the class directly? That way we would also exercise the class dispatch, and if `forgeVote` were ever refactored these tests would catch regressions.

(Not blocking — the implementations are identical today.)

**Review by dnadales — COMMENTED** [review]
Than you for adding the tests! Only one substantial concern about the use of `choose` and some minor observations.

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/Utils.hs (hunk 4 above)
Comment (agustinmista) [reply ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/Utils.hs:93]: Good catch! :sweat_smile:

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs (hunk 5 above)
Comment (agustinmista) [reply ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs]: Great idea! I don't remember exactly why I had the `castVote` helpers originally, but it no longer makes sense to have them. I refactored the code to call `forgeVote` from the class directly :)

File: ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs (hunk 1 above)
Comment (agustinmista) [reply ouroboros-consensus/test/consensus-test/Test/Consensus/Committee/WFALS/Tests.hs:378]: Good idea! 

Even if we don't hit the corner case in practice (which I don't think we do because it would make the property fail), having the guards is good for documentation :+1:

**Review by dnadales — APPROVED** [review]
Thanks for incorporating the changes 👍
